histomicstk/features/FeatureExtraction.py

```diff
@@ -73,374 +78,98 @@ def FeatureExtraction(Label, In, Ic, K=128, Fs=6, Delta=8):
-
-    # calculate hematoxlyin features, capture feature names
-    HematoxylinIntensityGroup = IntensityFeatureGroup(In, Nuclei)
-    HematoxylinTextureGroup = TextureFeatureGroup(In, Nuclei)
-    HematoxylinGradientGroup = GradientFeatureGroup(In, Nuclei)
-    # calculate eosin features
-    EosinIntensityGroup = IntensityFeatureGroup(Ic, Cytoplasms)
-    EosinTextureGroup = TextureFeatureGroup(Ic, Cytoplasms)
-    EosinGradientGroup = GradientFeatureGroup(Ic, Cytoplasms)
+            np.logical_xor(Mask, binary_dilation(Nucleus, Disk))
+        )
 
-    # add columns to dataframe
-    df['X'] = CentroidX
-    df['Y'] = CentroidY
+        # get list of cytoplasm pixels
+        regionCoords = np.argwhere(cytoplasm == 1)
+        regionCoords[:, 0] = regionCoords[:, 0] + min_row
+        regionCoords[:, 1] = regionCoords[:, 1] + min_col
 
-    df['Area'] = Area
-    df['Perimeter'] = Perimeter
-    df['Eccentricity'] = Eccentricity
-    df['Circularity'] = Circularity
-    df['MajorAxisLength'] = MajorAxisLength
-    df['MinorAxisLength'] = MinorAxisLength
-    df['Extent'] = Extent
-    df['Solidity'] = Solidity
+        # compute Texture, Gradient, Intensity features
+        EosinGradientGroup[i, :] = \
+            ComputeGradientFeatures(Ic, regionCoords, diffGc, BW_cannyc)
 
-    for i in range(0, Fs):
-        df['FSD' + str(i+1)] = FSDGroup[:, i]
+    # initialize panda dataframe
+    df = pd.DataFrame()
 
-    for f in HematoxylinIntensityGroup._fields:
-        df['Hematoxylin' + f] = getattr(HematoxylinIntensityGroup, f)
+    fmorph = ComputeMorphometryFeatures(Label, rprops=regions)
+    df = pd.concat([df, fmorph], axis=1)
 
-    for f in HematoxylinTextureGroup._fields:
-        df['Hematoxylin' + f] = getattr(HematoxylinTextureGroup, f)
+    ffsds = ComputeFSDs(Label, K, Fs, Delta, rprops=regions)
+    df = pd.concat([df, ffsds], axis=1)
 
-    for f in HematoxylinGradientGroup._fields:
-        df['Hematoxylin' + f] = getattr(HematoxylinGradientGroup, f)
+    GradientNames = ['MeanGradMag', 'StdGradMag', 'EntropyGradMag',
+                     'EnergyGradMag', 'SkewnessGradMag', 'KurtosisGradMag',
+                     'SumCanny', 'MeanCanny']
 
-    for f in EosinIntensityGroup._fields:
-        df['Cytoplasm' + f] = getattr(EosinIntensityGroup, f)
+    for i in range(0, len(GradientNames)):
+        df['Hematoxylin' + GradientNames[i]] = \
+            HematoxylinGradientGroup[:, i]
+        df['Cytoplasm' + GradientNames[i]] = EosinGradientGroup[:, i]
 
-    for f in EosinTextureGroup._fields:
-        df['Cytoplasm' + f] = getattr(EosinTextureGroup, f)
+    fint_nuclei = ComputeIntensityFeatures(Label, In, rprops=regions)
+    fint_nuclei.columns = ['Nucleus' + col for col in fint_nuclei.columns]
+    df = pd.concat([df, fint_nuclei], axis=1)
 
-    for f in EosinGradientGroup._fields:
-        df['Cytoplasm' + f] = getattr(EosinGradientGroup, f)
+    fint_cytoplasm = ComputeIntensityFeatures(Label, Ic, rprops=regions)
```

Comment:
We didn't really talk about this yet - but the basic idea is to 1. dilate each nucleus 2. remove pixels corresponding to any nucleus from this dilated mask and 3. calculate statistics from the eosin channel (from color deconvolution) in this region.

This also brings up the point of how we will handle other images besides hematoxylin and eosin. The feature extraction will eventually need to be aware of what stains are coming, and what the expected localization is (nuclear, membrane, non-nuclear or some combination).